crates/hir-expand/src/cfg_process.rs

```diff
@@ -0,0 +1,204 @@
+//! Processes out #[cfg] and #[cfg_attr] attributes from the input for the derive macro
+use rustc_hash::FxHashSet;
+use syntax::{
+    ast::{self, Attr, HasAttrs, Meta, VariantList},
+    AstNode, SyntaxElement, SyntaxNode, T,
+};
+use tracing::{debug, warn};
+
+use crate::{db::ExpandDatabase, MacroCallKind, MacroCallLoc};
+
+fn check_cfg_attr(attr: &Attr, loc: &MacroCallLoc, db: &dyn ExpandDatabase) -> Option<bool> {
+    if !attr.simple_name().as_deref().map(|v| v == "cfg")? {
+        return None;
+    }
+    debug!("Evaluating cfg {}", attr);
+    let cfg = cfg::CfgExpr::parse_from_attr_meta(attr.meta()?)?;
+    debug!("Checking cfg {:?}", cfg);
+    let enabled = db.crate_graph()[loc.krate].cfg_options.check(&cfg) != Some(false);
+    Some(enabled)
+}
+
+fn check_cfg_attr_attr(attr: &Attr, loc: &MacroCallLoc, db: &dyn ExpandDatabase) -> Option<bool> {
+    if !attr.simple_name().as_deref().map(|v| v == "cfg_attr")? {
+        return None;
+    }
+    debug!("Evaluating cfg_attr {}", attr);
+    let cfg_expr = cfg::CfgExpr::parse_from_attr_meta(attr.meta()?)?;
+    debug!("Checking cfg_attr {:?}", cfg_expr);
+    let enabled = db.crate_graph()[loc.krate].cfg_options.check(&cfg_expr) != Some(false);
+    Some(enabled)
+}
+
+fn process_has_attrs_with_possible_comma<I: HasAttrs>(
+    items: impl Iterator<Item = I>,
+    loc: &MacroCallLoc,
+    db: &dyn ExpandDatabase,
+    remove: &mut FxHashSet<SyntaxElement>,
+) -> Option<()> {
+    for item in items {
+        let field_attrs = item.attrs();
+        'attrs: for attr in field_attrs {
+            if check_cfg_attr(&attr, loc, db).map(|enabled| !enabled).unwrap_or_default() {
+                debug!("censoring type {:?}", item.syntax());
+                remove.insert(item.syntax().clone().into());
+                // We need to remove the , as well
+                add_comma(&item, remove);
+                break 'attrs;
+            }
+
+            if let Some(enabled) = check_cfg_attr_attr(&attr, loc, db) {
+                if enabled {
+                    debug!("Removing cfg_attr tokens {:?}", attr);
+                    let meta = attr.meta()?;
+                    let removes_from_cfg_attr = remove_tokens_within_cfg_attr(meta)?;
+                    remove.extend(removes_from_cfg_attr);
+                } else {
+                    debug!("censoring type cfg_attr {:?}", item.syntax());
+                    remove.insert(attr.syntax().clone().into());
+                    continue;
+                }
+            }
+        }
+    }
+    Some(())
+}
+
+fn remove_tokens_within_cfg_attr(meta: Meta) -> Option<FxHashSet<SyntaxElement>> {
```

Comment:
We should probably remove all tokens after the initial subtree that we wanna keep post expansion (that is `#[cfg_attr(foo, bar, baz)]` should be censored into `#[bar]`, instead of `#[bar, baz]` while still incorrect, that probably prevents more issues from invalid attributes being passed to proc-macros.)

Reply:
So I was going to handle that case next. https://github.com/rust-lang/rust-analyzer/issues/10110

The reason I did not was I am needing to append tokens not just remove them. Because to do that properly I would need to create a second seperate attribute. I know this is a bug. However, currently `syntax_node_to_token_tree_modified`
The `append` input is a `FxHashMap<SyntaxElement, Vec<tt::Leaf<SpanData<Ctx>>>>` and I am not sure however to convert a full attribute into that. So I was going to let that be a bug on its own. When I tested this case. 

What we see is a unexpected comma. I would be happy to fix this in this PR. However, I would need some guidance on appending a new attribute onto the code.

Reply:
I was mainly saying to do damage mitigation until we fully implement this, as syn proc-macros will just error out if we give them a malformed attribute  over one that is just missing information. We need to adjust `append`'s type here, since attributes consist of token trees and we are unable to model those with just leaves. We used to have `Subtree`s in there before but I removed that due to the unnecessary delimiters created. I think we want to change that type into `FxHashMap<SyntaxElement, Vec<Either<tt::Leaf<SpanData<Ctx>>, tt::Subtree<SpanData<Ctx>>>>>` or similar. A bit ugly but that is the most permiissible that should support he use case needed here (can re-evaluate after we see whats needed for the implementation). Then we can represent an attribute with a `#` leaf punc node and a `[...]` subtree